File: src/parser.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct CodeOwnerRule {
    pub pattern: String, // Normalized pattern (no leading/trailing /)
    pub owners: Vec<String>,
    pub original_path: String, // Original path from file (with / if present)
}

pub struct InvalidLine {
    pub line_number: usize,
    pub content: String,
}

pub fn parse_codeowners_file(
    file_path: &str,
) -> io::Result<(Vec<CodeOwnerRule>, Vec<InvalidLine>)> {
    let file = File::open(file_path)?;
    let reader = BufReader::with_capacity(64 * 1024, file);

    let mut rules = Vec::with_capacity(1000);
    let mut invalid_lines = Vec::new();

    for (line_number, line_result) in reader.lines().enumerate() {
        if let Ok(line) = line_result {
            let trimmed_line = line.trim();
            if trimmed_line.is_empty() || trimmed_line.starts_with('#') {
                // Skip empty lines and comments
                continue;
            }

            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.is_empty() {
                continue;
            }

            let pattern = parts[0].trim_matches('/').to_string();
            let owners = parts[1..].iter().map(|s| s.to_string()).collect();
            let original_path = parts[0].to_string();

            // Basic validation - ensure pattern is not empty after trimming
            if pattern.is_empty() {
                let invalid_line = InvalidLine {
                    line_number: line_number + 1,
                    content: line,
                };
                invalid_lines.push(invalid_line);
                continue;
            }

            // Check for invalid glob patterns
            if validate_pattern(&pattern, &original_path).is_err() {
                let invalid_line = InvalidLine {
                    line_number: line_number + 1,
                    content: line,
                };
                invalid_lines.push(invalid_line);
                continue;
            }

            let rule = CodeOwnerRule {
                pattern,
                owners,
                original_path,
            };

            rules.push(rule);
        }
    }

    rules.shrink_to_fit();
    invalid_lines.shrink_to_fit();

    Ok((rules, invalid_lines))
}

// Validate that the pattern can be turned into valid globs
fn validate_pattern(pattern: &str, original_path: &str) -> Result<(), &'static str> {
    use globset::Glob;

    let is_anchored = original_path.starts_with('/');
    let is_directory = original_path.ends_with('/');

    // Test that we can create the necessary globs
    match (is_anchored, is_directory) {
        (true, true) => {
            // /docs/ → need to create "docs" and "docs/**"
            Glob::new(pattern).map_err(|_| "invalid pattern")?;
            Glob::new(&format!("{}/**", pattern)).map_err(|_| "invalid pattern")?;
        }
        (true, false) => {
            // /src/file.rs → need to create "src/file.rs"
            Glob::new(pattern).map_err(|_| "invalid pattern")?;
        }
        (false, true) => {
            // lib/ → need to create "**/lib" and "**/lib/**"
            Glob::new(&format!("**/{}", pattern)).map_err(|_| "invalid pattern")?;
            Glob::new(&format!("**/{}/**", pattern)).map_err(|_| "invalid pattern")?;
        }
        (false, false) => {
            // *.rs or file.txt → need to create pattern or "**/pattern"
            if pattern.contains('*') || pattern.contains('?') || pattern.contains('[') {
                Glob::new(pattern).map_err(|_| "invalid pattern")?;
            } else {
                Glob::new(&format!("**/{}", pattern)).map_err(|_| "invalid pattern")?;
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::write;
    use tempfile::NamedTempFile;

    fn with_temp_codeowners(content: &str) -> NamedTempFile {
        let file = NamedTempFile::new().unwrap();
        write(file.path(), content).unwrap();
        file
    }

    #[test]
    fn parses_valid_lines() {
        let file = with_temp_codeowners("src/lib.rs @alice\n");
        let (rules, invalids) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules.len(), 1);
        assert_eq!(rules[0].pattern, "src/lib.rs");
        assert_eq!(rules[0].owners, vec!["@alice"]);
        assert_eq!(rules[0].original_path, "src/lib.rs");
        assert!(invalids.is_empty());
    }

    #[test]
    fn ignores_comments_and_blanks() {
        let file = with_temp_codeowners("# comment\n\nsrc/main.rs @bob\n");
        let (rules, _) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules.len(), 1);
        assert_eq!(rules[0].owners, vec!["@bob"]);
    }

    #[test]
    fn detects_invalid_glob() {
        let file = with_temp_codeowners("docs/[ @bad\n");
        let (_, invalids) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(invalids.len(), 1);
        assert!(invalids[0].content.contains("docs/["));
    }

    #[test]
    fn trims_leading_trailing_slashes() {
        let file = with_temp_codeowners("/foo/ @team\n");
        let (rules, _) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules[0].pattern, "foo");
        assert_eq!(rules[0].original_path, "/foo/");
    }

    #[test]
    fn parses_multiple_owners() {
        let file = with_temp_codeowners("src/ @alice @bob\n");
        let (rules, _) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules[0].owners, vec!["@alice", "@bob"]);
    }

    #[test]
    fn handles_wildcard_patterns() {
        let file = with_temp_codeowners("*.md @docs-team\n**/*.rs @rust-team\n");
        let (rules, invalids) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules.len(), 2);
        assert!(invalids.is_empty());
        assert_eq!(rules[0].pattern, "*.md");
        assert_eq!(rules[1].pattern, "**/*.rs");
    }

    #[test]
    fn handles_anchored_patterns() {
        let file = with_temp_codeowners("/README.md @docs\n/src/ @dev\n");
        let (rules, _) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules.len(), 2);
        assert_eq!(rules[0].pattern, "README.md");
        assert_eq!(rules[0].original_path, "/README.md");
        assert_eq!(rules[1].pattern, "src");
        assert_eq!(rules[1].original_path, "/src/");
    }

    #[test]
    fn rejects_empty_pattern() {
        let file = with_temp_codeowners("/ @team\n");
        let (rules, invalids) = parse_codeowners_file(file.path().to_str().unwrap()).unwrap();
        assert_eq!(rules.len(), 0);
        assert_eq!(invalids.len(), 1);
    }
}

File: internal/parser/parser.go
package parser

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

// CodeOwnerRule represents a parsed CODEOWNERS rule
type CodeOwnerRule struct {
	Pattern      string   `json:"pattern"`      // Normalized pattern (no leading/trailing /)
	Owners       []string `json:"owners"`       // List of owners
	OriginalPath string   `json:"originalPath"` // Original path from file (with / if present)
}

// InvalidLine represents a line that couldn't be parsed
type InvalidLine struct {
	LineNumber int    `json:"lineNumber"`
	Content    string `json:"content"`
}

// ParseResult contains the parsing results
type ParseResult struct {
	Rules        []CodeOwnerRule `json:"rules"`
	InvalidLines []InvalidLine   `json:"invalidLines"`
}

// ParseCodeOwnersFile parses a CODEOWNERS file and returns rules and invalid lines
func ParseCodeOwnersFile(filePath string) (*ParseResult, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s: %w", filePath, err)
	}
	defer file.Close()

	return ParseCodeOwners(file)
}

// ParseCodeOwners parses CODEOWNERS content from a reader
func ParseCodeOwners(reader io.Reader) (*ParseResult, error) {
	// Count non-empty, non-comment lines first for accurate pre-allocation
	ruleCount := countRules(reader)

	// Reset reader to beginning
	if seeker, ok := reader.(io.Seeker); ok {
		seeker.Seek(0, io.SeekStart)
	} else {
		// If reader is not seekable, we'll fall back to default estimation
		if ruleCount == 0 {
			ruleCount = 1000 // Conservative estimate
		}
	}

	// Pre-allocate slices with exact capacity for optimal performance
	rules := make([]CodeOwnerRule, 0, ruleCount)
	invalidLines := make([]InvalidLine, 0, ruleCount/20) // Estimate ~5% invalid lines

	scanner := bufio.NewScanner(reader)
	lineNumber := 0

	for scanner.Scan() {
		lineNumber++
		line := scanner.Text()
		trimmedLine := strings.TrimSpace(line)

		// Skip empty lines and comments
		if trimmedLine == "" || strings.HasPrefix(trimmedLine, "#") {
			continue
		}

		parts := strings.Fields(line)
		if len(parts) == 0 {
			continue
		}

		originalPath := parts[0]
		pattern := strings.Trim(originalPath, "/")
		owners := parts[1:]

		// Basic validation - ensure pattern is not empty after trimming
		if pattern == "" {
			invalidLines = append(invalidLines, InvalidLine{
				LineNumber: lineNumber,
				Content:    line,
			})
			continue
		}

		// Validate the pattern can be used as a valid glob
		if err := validatePattern(pattern, originalPath); err != nil {
			invalidLines = append(invalidLines, InvalidLine{
				LineNumber: lineNumber,
				Content:    line,
			})
			continue
		}

		rule := CodeOwnerRule{
			Pattern:      pattern,
			Owners:       owners,
			OriginalPath: originalPath,
		}

		rules = append(rules, rule)
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("error reading file: %w", err)
	}

	return &ParseResult{
		Rules:        rules,
		InvalidLines: invalidLines,
	}, nil
}

// validatePattern checks if the pattern can be turned into valid globs
func validatePattern(pattern, originalPath string) error {
	isAnchored := strings.HasPrefix(originalPath, "/")
	isDirectory := strings.HasSuffix(originalPath, "/")

	// Test that we can create the necessary globs
	switch {
	case isAnchored && isDirectory:
		// /docs/ → need to create "docs" and "docs/**"
		if !doublestar.ValidatePattern(pattern) {
			return fmt.Errorf("invalid pattern: %s", pattern)
		}
		if !doublestar.ValidatePattern(pattern + "/**") {
			return fmt.Errorf("invalid pattern: %s/**", pattern)
		}
	case isAnchored && !isDirectory:
		// /src/file.rs → need to create "src/file.rs"
		if !doublestar.ValidatePattern(pattern) {
			return fmt.Errorf("invalid pattern: %s", pattern)
		}
	case !isAnchored && isDirectory:
		// lib/ → need to create "**/lib" and "**/lib/**"
		if !doublestar.ValidatePattern("**/" + pattern) {
			return fmt.Errorf("invalid pattern: **/%s", pattern)
		}
		if !doublestar.ValidatePattern("**/" + pattern + "/**") {
			return fmt.Errorf("invalid pattern: **/%s/**", pattern)
		}
	case !isAnchored && !isDirectory:
		// *.rs or file.txt → need to create pattern or "**/pattern"
		if strings.ContainsAny(pattern, "*?[]") {
			// Already a wildcard pattern like *.rs, **/*.md
			if !doublestar.ValidatePattern(pattern) {
				return fmt.Errorf("invalid pattern: %s", pattern)
			}
		} else {
			// Plain file like config.json → match "**/config.json"
			if !doublestar.ValidatePattern("**/" + pattern) {
				return fmt.Errorf("invalid pattern: **/%s", pattern)
			}
		}
	}

	return nil
}

// countRules counts the number of non-empty, non-comment lines for pre-allocation
func countRules(reader io.Reader) int {
	scanner := bufio.NewScanner(reader)
	count := 0

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		// Count potential rules (non-empty, non-comment lines)
		if line != "" && !strings.HasPrefix(line, "#") {
			count++
		}
	}

	return count
}
